File: RiptideCore/Sources/RiptideCore/ProgramGenerator.swift
import Foundation

public struct GeneratorInput: Sendable {
    public let effort: Effort
    public let days: Int
    /// Ordered as the user picked them; every present muscle has ≥1 exercise.
    public let selections: [MuscleGroup: [ExerciseDefinition]]
    public init(effort: Effort, days: Int, selections: [MuscleGroup: [ExerciseDefinition]]) {
        self.effort = effort
        self.days = days
        self.selections = selections
    }
}

public struct GeneratedLift: Equatable, Sendable {
    public let exercise: ExerciseDefinition
    public let sets: Int
}

public struct GeneratedDay: Equatable, Sendable {
    public let lifts: [GeneratedLift]
}

public struct GeneratedProgram: Equatable, Sendable {
    public let days: [GeneratedDay]
}

public enum ProgramGenerator {
    /// Deterministic, total over wizard-valid input (spec §5). Never throws.
    public static func generate(_ input: GeneratorInput) -> GeneratedProgram {
        precondition(input.effort.allowedDays.contains(input.days), "wizard gates day counts")

        let selected = MuscleGroup.processingOrder.filter { !(input.selections[$0] ?? []).isEmpty }
        var dayLifts: [[GeneratedLift]] = Array(repeating: [], count: input.days)
        var dayTotals = [Int](repeating: 0, count: input.days)
        var secondarySets: [MuscleGroup: Int] = [:]

        for (mIndex, muscle) in selected.enumerated() {
            let exercises = input.selections[muscle]!
            let range = VolumeTable.weeklyRange(for: muscle, effort: input.effort)
            var target = Allocation.weeklyTarget(range: range, days: input.days)

            // Spec §5.2: receivers get 0.5 credit per secondary set, floor, never below 0.
            if MuscleGroup.receivers.contains(muscle) {
                let credit = (secondarySets[muscle] ?? 0) / 2
                target = max(0, target - credit)
                if target == 1 { target = 2 } // min-appearance rule
            }
            guard target >= 2 else { continue }

            // Spec §5.6 ladder step 4: clamp to capacity. Capacity always covers
            // range.low for every reachable (muscle, effort, days, exerciseCount)
            // combination under the corrected volume table — see
            // testExhaustiveSweepFindsNoReachableShortfall in RiptideCoreTests.
            let capacity = input.days * exercises.count * 4
            let achieved = min(target, capacity)
            guard achieved >= 2 else { continue }

            let loads = Allocation.dayLoads(total: achieved, days: input.days, maxEntriesPerDay: exercises.count)

            // Spread this muscle's sessions evenly across the week so it isn't
            // trained several days in a row then left out (spec §5.5). Every
            // rotation of the even-spacing pattern is equally "spread", so pick
            // the rotation that best levels the running daily totals — this keeps
            // both goals: each muscle spaced out AND no day overloaded.
            let k = loads.count
            var dayOrder = Allocation.spreadDays(k: k, over: input.days, phase: 0)
                .sorted { dayTotals[$0] < dayTotals[$1] }
            var bestScore = Int.max
            for phase in 0..<input.days {
                let slots = Allocation.spreadDays(k: k, over: input.days, phase: phase)
                    .sorted { dayTotals[$0] < dayTotals[$1] }
                var trial = dayTotals
                for (i, load) in loads.enumerated() { trial[slots[i]] += load }
                let score = trial.reduce(0) { $0 + $1 * $1 }   // lower = more level
                if score < bestScore { bestScore = score; dayOrder = slots }
            }

            var rotation = 0
            for (i, load) in loads.enumerated() {
                let d = dayOrder[i]
                for size in Allocation.entrySizes(sets: load, maxEntries: exercises.count) {
                    let ex = exercises[rotation % exercises.count]
                    rotation += 1
                    dayLifts[d].append(GeneratedLift(exercise: ex, sets: size))
                    dayTotals[d] += size
                    for sec in ex.secondaries { secondarySets[sec, default: 0] += size }
                }
            }
        }

        // Within-day ordering: compounds-first display order (spec §5.7 / design PARTS order).
        let days = dayLifts.map { lifts in
            GeneratedDay(lifts: lifts.sorted { a, b in
                let ia = MuscleGroup.displayOrder.firstIndex(of: a.exercise.primary)!
                let ib = MuscleGroup.displayOrder.firstIndex(of: b.exercise.primary)!
                return ia == ib ? a.exercise.name < b.exercise.name : ia < ib
            })
        }
        return GeneratedProgram(days: days)
    }
}

File: web/src/core/programGenerator.ts
import type { Effort } from './effort';
import { allowedDays } from './effort';
import type { MuscleGroup } from './muscleGroup';
import { PROCESSING_ORDER, RECEIVERS, DISPLAY_ORDER } from './muscleGroup';
import type { ExerciseDefinition } from './exercise';
import { weeklyRange } from './volumeTable';
import { weeklyTarget, dayLoads, spreadDays, entrySizes } from './allocation';
import { idiv, assert } from './util';

export interface GeneratorInput {
  effort: Effort;
  days: number;
  // Ordered as the user picked them; every present muscle has >= 1 exercise.
  selections: Map<MuscleGroup, ExerciseDefinition[]>;
}
export interface GeneratedLift {
  exercise: ExerciseDefinition;
  sets: number;
}
export interface GeneratedDay {
  lifts: GeneratedLift[];
}
export interface GeneratedProgram {
  days: GeneratedDay[];
}

// Deterministic, total over wizard-valid input (spec §5). Never throws for valid input.
export function generate(input: GeneratorInput): GeneratedProgram {
  assert(allowedDays(input.effort).includes(input.days), 'wizard gates day counts');

  const selected = PROCESSING_ORDER.filter((m) => (input.selections.get(m) ?? []).length > 0);
  const dayLifts: GeneratedLift[][] = Array.from({ length: input.days }, () => []);
  const dayTotals: number[] = new Array(input.days).fill(0);
  const secondarySets = new Map<MuscleGroup, number>();

  for (const muscle of selected) {
    const exercises = input.selections.get(muscle)!;
    const range = weeklyRange(muscle, input.effort);
    let target = weeklyTarget(range);

    // Spec §5.2: receivers get 0.5 credit per secondary set. Aim at midpoint - credit,
    // but guarantee >= 2 direct sets whenever there's still room under the range top.
    if (RECEIVERS.includes(muscle)) {
      const credit = idiv(secondarySets.get(muscle) ?? 0, 2);
      const aim = Math.max(0, target - credit);
      const headroom = Math.max(0, range.high - credit);
      target = Math.max(aim, headroom >= 2 ? 2 : 0);
    }
    if (target < 2) continue;

    // Spec §5.6 ladder step 4: clamp to capacity.
    const capacity = input.days * exercises.length * 4;
    const achieved = Math.min(target, capacity);
    if (achieved < 2) continue;

    const loads = dayLoads(achieved, input.days, exercises.length);
    const k = loads.length;

    // Spread this muscle's sessions evenly; among equally-spread rotations, pick the
    // one that best levels running daily totals. Stable sort keeps spreadDays order on ties.
    let dayOrder = [...spreadDays(k, input.days, 0)].sort((a, b) => dayTotals[a] - dayTotals[b]);
    let bestScore = Number.MAX_SAFE_INTEGER;
    for (let phase = 0; phase < input.days; phase++) {
      const slots = [...spreadDays(k, input.days, phase)].sort((a, b) => dayTotals[a] - dayTotals[b]);
      const trial = [...dayTotals];
      loads.forEach((load, i) => {
        trial[slots[i]] += load;
      });
      const score = trial.reduce((s, v) => s + v * v, 0); // lower = more level
      if (score < bestScore) {
        bestScore = score;
        dayOrder = slots;
      }
    }

    let rotation = 0;
    loads.forEach((load, i) => {
      const d = dayOrder[i];
      for (const size of entrySizes(load, exercises.length)) {
        const ex = exercises[rotation % exercises.length];
        rotation += 1;
        dayLifts[d].push({ exercise: ex, sets: size });
        dayTotals[d] += size;
        for (const sec of ex.secondaries) {
          secondarySets.set(sec, (secondarySets.get(sec) ?? 0) + size);
        }
      }
    });
  }

  // Within-day ordering: compounds-first display order (spec §5.7). Stable sort keeps
  // insertion order for lifts with the same primary and name.
  const days: GeneratedDay[] = dayLifts.map((lifts) => ({
    lifts: [...lifts].sort((a, b) => {
      const ia = DISPLAY_ORDER.indexOf(a.exercise.primary);
      const ib = DISPLAY_ORDER.indexOf(b.exercise.primary);
      if (ia !== ib) return ia - ib;
      if (a.exercise.name < b.exercise.name) return -1;
      if (a.exercise.name > b.exercise.name) return 1;
      return 0;
    }),
  }));
  return { days };
}
